File: frontend/src/utils/validation.ts
import type { ValidationResult } from '../types';

const SUPPORTED_FILE_TYPES = [
  'application/pdf',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  'application/vnd.openxmlformats-officedocument.presentationml.presentation',
  'text/plain',
];

const SUPPORTED_EXTENSIONS = ['.pdf', '.docx', '.pptx', '.txt'];

const MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB

export const validateFile = (file: File, allowedTypes?: string[], maxSize?: number): ValidationResult => {
  const maxFileSize = maxSize || MAX_FILE_SIZE;
  const supportedTypes = allowedTypes || SUPPORTED_FILE_TYPES;
  
  // Check file size
  if (file.size > maxFileSize) {
    return {
      valid: false,
      error: `ファイルサイズが大きすぎます（最大${Math.round(maxFileSize / (1024 * 1024))}MB）`,
    };
  }

  // Check file type
  if (!supportedTypes.includes(file.type)) {
    // Also check by extension as fallback
    const extension = '.' + file.name.split('.').pop()?.toLowerCase();
    const supportedExtensions = allowedTypes ? 
      allowedTypes.map(type => {
        if (type.includes('pdf')) return '.pdf';
        if (type.includes('word') || type.includes('docx')) return '.docx';
        if (type.includes('presentation') || type.includes('pptx')) return '.pptx';
        if (type.includes('text')) return '.txt';
        return '';
      }).filter(Boolean) : 
      SUPPORTED_EXTENSIONS;
      
    if (!supportedExtensions.includes(extension)) {
      return {
        valid: false,
        error: 'サポートされていないファイル形式です（PDF、DOCX、PPTX、TXTのみ）',
      };
    }
  }

  return { valid: true };
};

export const validateFiles = (files: File[], allowedTypes?: string[], maxSize?: number): ValidationResult => {
  for (const file of files) {
    const result = validateFile(file, allowedTypes, maxSize);
    if (!result.valid) {
      return result;
    }
  }
  return { valid: true };
};

export const sanitizeQuery = (query: string): string => {
  return query.trim().replace(/[<>]/g, '');
};

export const validateQuery = (query: string): ValidationResult => {
  const sanitized = sanitizeQuery(query);
  
  if (!sanitized) {
    return {
      valid: false,
      error: '質問を入力してください',
    };
  }

  if (sanitized.length < 3) {
    return {
      valid: false,
      error: '質問は3文字以上入力してください',
    };
  }

  if (sanitized.length > 500) {
    return {
      valid: false,
      error: '質問は500文字以内で入力してください',
    };
  }

  return { valid: true };
};

export const formatFileSize = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';

  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};

export const getFileExtension = (fileName: string): string => {
  return '.' + fileName.split('.').pop()?.toLowerCase() || '';
};

export const getFileTypeIcon = (fileType: string): string => {
  if (fileType.includes('pdf')) return '📄';
  if (fileType.includes('word') || fileType.includes('docx')) return '📝';
  if (fileType.includes('presentation') || fileType.includes('pptx')) return '📊';
  if (fileType.includes('text')) return '📃';
  return '📎';
};

File: internal/utils/validation.go
package utils

import (
	"mime"
	"path/filepath"
	"strings"

	"aws-serverless-rag/internal/models"
)

// ValidateFileUpload validates file upload parameters
func ValidateFileUpload(filename, contentType string, fileSize int64) error {
	// Check file name
	if filename == "" {
		return models.ErrInvalidRequest
	}

	// Check file extension
	ext := strings.ToLower(filepath.Ext(filename))
	if !isValidFileExtension(ext) {
		return models.ErrInvalidFileType
	}

	// Check content type
	if contentType != "" && !models.IsValidFileType(contentType) {
		return models.ErrInvalidFileType
	}

	// Check file size (max 50MB)
	const maxFileSize = 50 * 1024 * 1024 // 50MB
	if fileSize > maxFileSize {
		return models.ErrInvalidRequest
	}

	return nil
}

// isValidFileExtension checks if the file extension is supported
func isValidFileExtension(ext string) bool {
	validExtensions := map[string]bool{
		".pdf":  true,
		".doc":  true,
		".docx": true,
		".ppt":  true,
		".pptx": true,
		".txt":  true,
		".md":   true,
	}
	return validExtensions[ext]
}

// DetectContentType detects content type from file extension
func DetectContentType(filename string) string {
	ext := strings.ToLower(filepath.Ext(filename))

	contentTypes := map[string]string{
		".pdf":  "application/pdf",
		".doc":  "application/msword",
		".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
		".ppt":  "application/vnd.ms-powerpoint",
		".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
		".txt":  "text/plain",
		".md":   "text/markdown",
	}

	if contentType, exists := contentTypes[ext]; exists {
		return contentType
	}

	// Fallback to mime detection
	return mime.TypeByExtension(ext)
}

// SanitizeFilename sanitizes the filename for safe storage
func SanitizeFilename(filename string) string {
	// Remove path components
	filename = filepath.Base(filename)

	// Replace spaces with underscores
	filename = strings.ReplaceAll(filename, " ", "_")

	// Remove any potentially dangerous characters
	filename = strings.Map(func(r rune) rune {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') ||
			(r >= '0' && r <= '9') || r == '.' || r == '_' || r == '-' {
			return r
		}
		return '_'
	}, filename)

	return filename
}

// ValidateQuery validates query parameters
func ValidateQuery(question string, maxResults int) error {
	if strings.TrimSpace(question) == "" {
		return models.ErrInvalidQuestion
	}

	if len(question) > 1000 {
		return models.ErrInvalidRequest
	}

	if maxResults < 1 || maxResults > 20 {
		return models.ErrInvalidRequest
	}

	return nil
}
